RPM dependencies for foreman_google package

stejskalleos opened the pull request:
For https://github.com/theforeman/foreman-packaging/pull/7777

Adds:
* `google-cloud-compute`
* `google-cloud-core`
* `google-cloud-compute-v1`
* `google-cloud-common`
* `google-cloud-errors`
* `gapic-common`
* `google-protobuf`
* `googleapis-common-protos`
* `googleapis-common-protos-types`

ekohl commented on packages/foreman/rubygem-google-cloud-compute/rubygem-google-cloud-compute.spec:
```diff
@@ -0,0 +1,64 @@
+%files
+%dir %{gem_instdir}
+%exclude %{gem_instdir}/.yardopts
+%{gem_instdir}/AUTHENTICATION.md
```
I think this is a doc file. You can probably also move it to `%files doc`:
```suggestion
%doc %{gem_instdir}/AUTHENTICATION.md
```

ekohl commented on packages/foreman/rubygem-google-cloud-compute/rubygem-google-cloud-compute.spec:
```diff
@@ -0,0 +1,64 @@
+%changelog
+* Mon Jul 18 2022 Leos Stejskal <lstejska@redhat.com> 0.4.0-1
+- Add rubygem-google-cloud-compute generated by gem2rpm using the default template
+
```
```suggestion
```

ekohl reviewed (commented):
Minor stylistic things, otherwise :+1:

ekohl commented:
If https://github.com/theforeman/foreman_google/pull/50 is merged and released, I believe this PR is no longer needed.

ekohl commented:
I started to look into this and we'll hit a problem with Faraday. [gapic-common](https://rubygems.org/gems/gapic-common) requires Faraday >= 1.9 and Katello (and Pulp plugins) require < 1.9.

stejskalleos commented:
Is there any update on the `Faraday` update?

ekohl commented:
Yes, https://ci.theforeman.org/job/katello-nightly-rpm-pipeline/1421/ is running now to verify Pulp can run with the updated gems. After that I think we can bump Faraday, but that's easier after that Katello run has completed.

stejskalleos commented:
[koji/rubygem-google-protobuf](http://koji.katello.org/koji/getfile?taskID=532750&volume=DEFAULT&name=build.log&offset=-4000) failed with:
```
error: File not found: /builddir/build/BUILDROOT/rubygem-google-protobuf-3.21.2-1.el8.x86_64/usr/lib64/gems/ruby/google-protobuf-3.21.2
Processing files: rubygem-google-protobuf-3.21.2-1.el8.x86_64
RPM build errors:
    File not found: /builddir/build/BUILDROOT/rubygem-google-protobuf-3.21.2-1.el8.x86_64/usr/lib64/gems/ruby/google-protobuf-3.21.2
Child return code was: 1
EXCEPTION: [Error()]
```

But gem is added in the commit, @theforeman/packaging any ideas how to fix the issue?

ehelms commented:
Try this:

```
diff --git a/packages/foreman/rubygem-google-protobuf/rubygem-google-protobuf.spec b/packages/foreman/rubygem-google-protobuf/rubygem-google-protobuf.spec
index 850548904..22c673959 100644
--- a/packages/foreman/rubygem-google-protobuf/rubygem-google-protobuf.spec
+++ b/packages/foreman/rubygem-google-protobuf/rubygem-google-protobuf.spec
@@ -14,12 +14,6 @@ BuildRequires: ruby-devel >= 2.3
 # Compiler is required for build of gem binary extension.
 # https://fedoraproject.org/wiki/Packaging:C_and_C++#BuildRequires_and_Requires
 BuildRequires: gcc
-# BuildRequires: rubygem(rake-compiler-dock) = 1.2.1
-# BuildRequires: rubygem(rake-compiler) >= 1.1.0
-# BuildRequires: rubygem(rake-compiler) < 1.2
-# BuildRequires: rubygem(test-unit) >= 3.0
-# BuildRequires: rubygem(test-unit) < 4
-# BuildRequires: rubygem(test-unit) >= 3.0.9
 
 %description
 Protocol Buffers are Google's data interchange format.
@@ -50,18 +44,13 @@ cp -a .%{gem_dir}/* \
         %{buildroot}%{gem_dir}/
 
 # This doesn't work
-# mkdir -p %{buildroot}%{gem_extdir_mri}
-# cp -a .%{gem_extdir_mri}/{gem.build_complete,*.so} %{buildroot}%{gem_extdir_mri}/
+mkdir -p %{buildroot}%{gem_extdir_mri}/google
+cp -a .%{gem_extdir_mri}/gem.build_complete %{buildroot}%{gem_extdir_mri}/
+cp -a .%{gem_extdir_mri}/google/*.so %{buildroot}%{gem_extdir_mri}/google/
 
 # Prevent dangling symlink in -debuginfo (rhbz#878863).
 rm -rf %{buildroot}%{gem_instdir}/ext/
 
-
-%check
-pushd .%{gem_instdir}
-# ruby -e 'Dir.glob "./test/**/*_test.rb", &method(:require)'
-popd
-
 %files
 %dir %{gem_instdir}
 %{gem_extdir_mri}
```

ehelms commented:
With the above change I got all building:

```
PLAY RECAP *******************************************************************************************************************************************************************************************************************************************************************************
rubygem-gapic-common       : ok=20   changed=3    unreachable=0    failed=0    skipped=1    rescued=0    ignored=1   
rubygem-google-cloud-common : ok=19   changed=3    unreachable=0    failed=0    skipped=1    rescued=0    ignored=1   
rubygem-google-cloud-compute : ok=19   changed=3    unreachable=0    failed=0    skipped=1    rescued=0    ignored=1   
rubygem-google-cloud-compute-v1 : ok=19   changed=3    unreachable=0    failed=0    skipped=1    rescued=0    ignored=1   
rubygem-google-cloud-core  : ok=19   changed=3    unreachable=0    failed=0    skipped=1    rescued=0    ignored=1   
rubygem-google-cloud-errors : ok=19   changed=3    unreachable=0    failed=0    skipped=1    rescued=0    ignored=1   
rubygem-google-protobuf    : ok=16   changed=1    unreachable=0    failed=0    skipped=4    rescued=0    ignored=0   
rubygem-googleapis-common-protos : ok=19   changed=3    unreachable=0    failed=0    skipped=1    rescued=0    ignored=1   
rubygem-googleapis-common-protos-types : ok=19   changed=3    unreachable=0    failed=0    skipped=1    rescued=0    ignored=1   
```

ehelms commented:
Also you can drop -- `rel-eng/tito.props`

stejskalleos commented:
@ehelms I updated the `rubygem-google-protobuf.spec` file and removed `rel-eng/tito.props`, build continued until it failed on missing [grpc gem](https://rubygems.org/gems/grpc/versions/1.48.0). So I added but now it's failing on `obal lint: error: argument target: invalid choice: 'rubygem-grpc'`

`invalid choice`, what does it mean?

ehelms commented:
> @ehelms I updated the `rubygem-google-protobuf.spec` file and removed `rel-eng/tito.props`, build continued until it failed on missing [grpc gem](https://rubygems.org/gems/grpc/versions/1.48.0). So I added but now it's failing on `obal lint: error: argument target: invalid choice: 'rubygem-grpc'`
> 
> `invalid choice`, what does it mean?

That means that it could not find `rubygem-grpc` listed in the Ansible inventory that Obal uses. That inventory is derived from the package_manifest.yml. So you just need to add rubygem-grpc to the package_manifest.

ehelms commented:
There are build dependencies in this PR needed for `rubygem-grpc` so you'll want to break that out to it's own PR to be merged after this PR:

```
DEBUG util.py:634:  No matching package to install: 'rubygem(google-protobuf) >= 3.19'
DEBUG util.py:634:  No matching package to install: 'rubygem(google-protobuf) < 4'
DEBUG util.py:634:  No matching package to install: 'rubygem(googleapis-common-protos-types) >= 1.0'
DEBUG util.py:634:  No matching package to install: 'rubygem(googleapis-common-protos-types) < 2'
DEBUG util.py:634:  Not all dependencies satisfied
```

ehelms commented:
> @ehelms but why it's complaining about packages that are added in this PR? I mean even the versions are correct:
> 
> * `rubygem-google-protobuf-3.21.2-1.src.rpm`
> * `rubygem-googleapis-common-protos-types-1.3.2-1.src.rpm`

These are added here, but they are not built yet and available in the buildroot. `BuildRequires` must get merged, and built ahead of time in Koji so that they are available for installation inside Koji. `Requires` are only used at runtime and can be in the same PR.

So for rubygem-grpc (https://github.com/theforeman/foreman-packaging/pull/8248/files#diff-e5514bd3cbe49753335eca4852dd52340241257db49ebbed228fe09afa6f5dadR17-R20) those have to get built separately from the package itself, and first.

ehelms commented:
Extracted googleapis-common-protos-types here -- https://github.com/theforeman/foreman-packaging/pull/8548

ehelms commented:
As I have broken a few things out from this PR, the dependencies mainly, I also re-did the intent of this within the plugins directory context and opened a PR -- https://github.com/theforeman/foreman-packaging/pull/8549

ehelms commented:
Closing in favor of https://github.com/theforeman/foreman-packaging/pull/8549 -- thanks for the initial work @stejskalleos
